### PR TITLE
fcntl adding apple F_PREALLOCATE flag.

### PR DESCRIPTION
**Pull request by devnexen**
## Checklist:

- [x] I have read `CONTRIBUTING.md`
- [x] I have written necessary tests and rustdoc comments
- [x] A change log has been added if this PR modifies nix's API

### CODE REVIEW

**Comment by SteveLauC** [comment]
I am working on the CI issues, they are not related to this PR

**Comment by SteveLauC** [comment]
Gentle ping on this PR, would you like to rebase and finish it?

**Comment by devnexen** [comment]
seems libc 0.2.154 not yet available (just the tag on the repo but not yet published maybe) ?

**Comment by SteveLauC** [comment]
> seems libc 0.2.154 not yet available (just the tag on the repo but not yet published maybe) ?

Right, it has been yanked, we have to wait for https://github.com/rust-lang/libc/pull/3682

**Comment by SteveLauC** [comment]
The issue has been fixed, please rebase your branch:)

File: src/fcntl.rs [hunk 1]
```diff
@@ -692,6 +692,9 @@ pub enum FcntlArg<'a> {
     /// Return the full path without firmlinks of the fd.
     #[cfg(apple_targets)]
     F_GETPATH_NOFIRMLINK(&'a mut PathBuf),
+    /// Pre-allocate storage with different policies on fd.
+    #[cfg(apple_targets)]
+    F_PREALLOCATE(&'a libc::fstore_t),
```
Comment (SteveLauC) [inline src/fcntl.rs:697]: The associated buffer has an `OUT` field: 

```c
typedef struct fstore {
	    u_int32_t fst_flags;      /* IN: flags word */
	    int       fst_posmode;    /* IN: indicates offset field */
	    off_t     fst_offset;     /* IN: start of the region */
	    off_t     fst_length;     /* IN: size of the region */
	    off_t     fst_bytesalloc; /* OUT: number of bytes allocated */
} fstore_t;
```

i.e., the buffer can be written, if we use an immutable reference here, then UB could happen


Ref: https://keith.github.io/xcode-man-pages/fcntl.2.html#F_PREALLOCATE

Comment (SteveLauC) [reply src/fcntl.rs:697]: I just tried implementing this on my mac, I think we should create a helper type for those `IN` fields, and return the `OUT` field through `fcntl()`.

For that `fst_falgs` field, we would create a bitflags struct, for `fst_posmode`, a Rust enum would suffice.


And [F_ALLOCATEPERSIST](https://keith.github.io/xcode-man-pages/fcntl.2.html#F_ALLOCATEPERSIST) is not available in libc, we should add it first.

**Review by SteveLauC — CHANGES_REQUESTED** [review]

File: src/fcntl.rs (hunk 1 above)
Comment (SteveLauC) [reply src/fcntl.rs:697]: libc PR filed: https://github.com/rust-lang/libc/pull/3713

File: src/fcntl.rs [hunk 2]
```diff
@@ -800,6 +800,9 @@ pub enum FcntlArg<'a> {
     /// Issue an advisory read async with no copy to user
     #[cfg(apple_targets)]
     F_RDADVISE(libc::radvisory),
+    /// Pre-allocate storage with different policies on fd.
```
Comment (SteveLauC) [inline src/fcntl.rs:803]: Do you think we shoud document the reason why a mutable reference is needed here? i.e., there is a OUT field in the associated sturcutre🤔

**Review by SteveLauC — APPROVED** [review]
LGTM, thanks!

**Comment by SteveLauC** [comment]
Looks like you need to resolve the conflict in `src/fcntl.rs`.

**Pull request merged**